lib/src/java/facter.cc

```diff
@@ -0,0 +1,102 @@
+#include "com_puppetlabs_Facter.h"
+#include <facter/facts/collection.hpp>
+#include <facter/facts/scalar_value.hpp>
+#include <facter/facts/array_value.hpp>
+#include <facter/facts/map_value.hpp>
+#include <string>
+
+using namespace std;
+using namespace facter::facts;
+
+static string to_string(JNIEnv* env, jstring str)
+{
+    auto ptr = env->GetStringUTFChars(str, nullptr);
+    if (!ptr) {
+        return {};
+    }
+
+    auto size = env->GetStringUTFLength(str);
+    string result(ptr, size);
+
+    env->ReleaseStringUTFChars(str, ptr);
+    return result;
+}
+
+static jobject to_object(JNIEnv* env, char const* str)
+{
+    return env->NewStringUTF(str);
+}
+
+static jobject to_object(JNIEnv* env, int64_t value)
+{
+    jclass cls = env->FindClass("java/lang/Integer");
+    jmethodID methodID = env->GetMethodID(cls, "<init>", "(I)V");
```

Comment:
FindClass and GetMethodID can be expense, so depending on how frequently jruby server calls into jni, we may want to cache these, same comment below too.